### PR TITLE
[Merged by Bors] - feat(Topology): the partition of a space into fibers of a function

### PR DESCRIPTION
**Pull request by dagurtomas**
---

On its own this PR may seem strange, but this API is useful in #15321 to prove that discrete condensed sets are given by locally constant maps.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Logic/Function/FiberPartition.lean [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+set, and the forgetful functor from the category of condensed sets to the category of sets
+(see PR #14027).
+-/
+
+variable {X Y Z : Type*} (f : Y → Z) (g : X → Y)
+
+namespace Function
+
+/-- The indexing set of the partition. -/
+def Fibers : Type _ := Set.range (fun (x : Set.range f) ↦ f ⁻¹' {x.val})
```
Comment (jcommelin) [inline Mathlib/Logic/Function/FiberPartition.lean:23]: This won't work with the new `variable` algorithm. Please copy `f` syntax-explicitly into the statement:
```suggestion
def Fibers (f : Y → Z) : Type _ := Set.range (fun (x : Set.range f) ↦ f ⁻¹' {x.val})
```

File: Mathlib/Logic/Function/FiberPartition.lean [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  convert x.prop
+
+lemma mk_image (y : Y) : (Fibers.mk f y).image = f y :=
+  (map_eq_image (x := mkSelf f y)).symm
+
+lemma mem_iff_eq_image (y : Y) (a : Fibers f) : y ∈ a.val ↔ f y = a.image :=
+  ⟨fun h ↦ a.map_eq_image _ ⟨y, h⟩, fun h ↦ by rw [a.eq_fiber_image]; exact h⟩
+
+/-- An arbitrary element of `a : Fibers f`. -/
+noncomputable def preimage (a : Fibers f) : Y := a.2.choose.2.choose
```
Comment (jcommelin) [inline Mathlib/Logic/Function/FiberPartition.lean:55]: Can you please add a proof of `Set.Nonempty a.1`?

File: Mathlib/Logic/Function/FiberPartition.lean [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+lemma map_preimage_eq_image (a : Fibers f) : f a.preimage = a.image := a.2.choose.2.choose_spec
```
Comment (jcommelin) [inline Mathlib/Logic/Function/FiberPartition.lean:57]: Should this be simp-lemma?

Comment (dagurtomas) [reply Mathlib/Logic/Function/FiberPartition.lean]: I don't think so. It's used once in each directions in the application, as far as I can tell

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**